### PR TITLE
Integrating Scala Parser Into Rust.

### PR DESCRIPTION
**Pull request by Josef-Vonasek**
### Pull Request Description

Somehow I managed to include the 200 commits from previous PR. Ignore them please.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md), [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: common/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/DocParser.scala [hunk 1]
```diff
@@ -276,53 +275,9 @@ object DocParserRunner {
   * Essentially it enables Doc Parser to create pretty HTML files from
   * documented code.
   *
-  * When Doc Parser finishes its job user can invoke DocParserHTMLGenerator by
-  * simply passing the output of Doc Parser onto function called
-  * [[DocParserHTMLGenerator.generateHTMLForEveryDocumented]], and it will
-  * automatically traverse through AST prepared by Doc Parser and generate
-  * HTML files in all appropriate places.
   */
 object DocParserHTMLGenerator {
```
[the remark this answers is not in the document]
Comment (mwu-tow) [reply common/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/DocParser.scala]: File IO breaks JS targeting I expect. Perhaps we could move the generator to a separate package?

[the remark this answers is not in the document]
Comment (Josef-Vonasek) [reply common/scala/syntax/specialization/shared/src/main/scala/org/enso/syntax/text/DocParser.scala]: As we have already discussed, file IO is not supposed to be in parser.

File: common/rust/parser/src/jsclient.rs [hunk 2]
```diff
@@ -1,30 +1,49 @@
...
+      Ok(Client {})
+   }
 }
 
 impl IsParser for Client {
-    fn parse(&mut self, _program: String) -> api::Result<api::AST> {
-        Err(api::interop_error(Error::NotImplemented))
-    }
+   fn parse(&mut self, _program: String) -> api::Result<api::AST> {
+      Ok(parse(_program))
```
[the remark this answers is not in the document]
Comment (Josef-Vonasek) [reply common/rust/parser/src/jsclient.rs]: Rust should be able to catch such exception.

File: common/rust/parser/build.rs [hunk 3]
```diff
@@ -0,0 +1,14 @@
+use std::fs::File;
+use std::io::prelude::*;
+
+const PARSER_PATH: &str =
```
[the remark this answers is not in the document]
Comment (Josef-Vonasek) [reply common/rust/parser/build.rs]: Done.

File: common/rust/parser/src/lib.rs [hunk 4]
```diff
@@ -26,21 +29,27 @@ impl Parser {
...
 }
 
 impl api::IsParser for Parser {
     fn parse(&mut self, program: String) -> api::Result<api::AST> {
         self.deref_mut().parse(program)
     }
 }
+
+#[wasm_bindgen]
+pub fn parse(s: String) -> String {
```
[the remark this answers is not in the document]
Comment (Josef-Vonasek) [reply common/rust/parser/src/lib.rs]: It's the function that is exposed to index.html

File: common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala [hunk 5]
```diff
@@ -5,8 +5,8 @@ import java.util.UUID
 import cats.{Foldable, Functor, Monoid}
 import cats.derived._
 import cats.implicits._
-import io.circe.Encoder
-import io.circe.Json
+  import io.circe.Encoder
+  import io.circe.Json
```
Comment (mwu-tow) [inline common/scala/syntax/definition/src/main/scala/org/enso/syntax/text/AST.scala]: Not intentional I think.

File: common/rust/parser/build.rs [hunk 6]
```diff
@@ -0,0 +1,21 @@
+use std::fs::File;
+use std::io::prelude::*;
+use std::io::BufReader;
+
+const PARSER_PATH: &str = "./pkg/scala-parser.js";
+const PARSER_PATH_FIX: &str = "./pkg/scala-parser-fix.js";
+
+fn main() -> std::io::Result<()> {
+    // FIX for a scalajs es-module bug (solved in scalajs 1.0)
+    let     file = File::open(PARSER_PATH)?;
```
[the remark this answers is not in the document]
Comment (Josef-Vonasek) [reply common/rust/parser/build.rs]: Interestingly, `#[cfg(target_arch = "wasm32")]` doesn't work here. I need to yet find out why.

File: common/rust/parser/README.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+You can run the tests by calling:
+```
+sbt syntaxJS/fullOptJS
+wasm-pack test --headless --chrome --firefox --safari common/rust/parser
```
Comment (mwu-tow) [inline common/rust/parser/README.md]: We should warn that due to an external issue "naive" run of tests in Chrome will yield an error, as per https://github.com/rustwasm/wasm-pack/issues/611 .

File: common/rust/parser/src/jsclient.rs [hunk 8]
```diff
@@ -1,30 +1,59 @@
...
+        Ok(Client {})
     }
 }
 
 impl IsParser for Client {
     fn parse(&mut self, _program: String) -> api::Result<api::Ast> {
-        Err(api::interop_error(Error::NotImplemented))
+        match parse(_program) {
+            Ok(json_ast) => Err(ParsingError(json_ast)),
+            Err(_) => api::interop_error(Error::ScalaException()),
```
Comment (mwu-tow) [inline common/rust/parser/src/jsclient.rs]: please align `=>`

File: common/rust/parser/src/lib.rs [hunk 9]
```diff
@@ -26,21 +31,22 @@ impl Parser {
...
-        Ok(parser)
+        Ok(Parser(parser))
     }
 
     /// Obtains a default parser implementation, panicking in case of failure.
     pub fn new_or_panic() -> Parser {
         Parser::new()
             .unwrap_or_else(|e| panic!("Failed to create a parser: {:?}", e))
     }
+
```
Comment (mwu-tow) [inline common/rust/parser/src/lib.rs]: No need for newline here.

**Pull request merged**